DBZ-4196 Support schema changes during incremental snapshot

kgalieva opened the pull request:
https://issues.redhat.com/browse/DBZ-4196

Blocked by https://issues.redhat.com/browse/DBZ-4197 to make the schema changes integration tests pass faster. Temporarily included DBZ-4197 related commit to this PR

#### Implementation: 

When incremental snapshot queries a database for the next chunk of rows, the rows have the table’s current schema. However, if the binlog stream is behind the in-memory representation the table’s schema may be different from the current schema. For connectors with historized schema the solution is to wait for the connector to receive the DDL event in the binlog stream and update the in-memory representation of the table’s schema. After that, the connector can use the cached table’s structure to produce the correct incremental snapshot chunk events.

In order to detect schema changes during the incremental snapshot, each chunk’s schema needs to be compared against the table’s schema saved in the incremental snapshot context. 

To capture the table’s schema at the beginning of the incremental snapshot the connector queries the database for the table’s schema between low and high watermarks just like it would query an incremental snapshot chunk. As soon as the connector gets a window close event from the binlog it means that the in-memory representation of the table’s schema is up to date with the schema that was captured at the beginning of the incremental snapshot. The connector queries the database one more time to verify that the table's schema didn’t change since the initial schema capture. If the schema has changed, then the process of capturing the incremental snapshot schema is repeated.
<img width="1209" alt="1" src="https://user-images.githubusercontent.com/553032/138471504-cb223299-b57f-4224-b796-8daf661ccfc2.png">

If the schema is the same then it gets marked as verified and is used for comparisons during the following chunks selections.

<img width="1178" alt="2" src="https://user-images.githubusercontent.com/553032/138533269-82dbc755-13f3-4a7e-86de-e669ac15df59.png">

When a chunk’s schema doesn’t match the schema saved in incremental snapshot context the connector drops the selected chunk, updates incremental snapshot schema in the snapshot context and re-reads the chunk after the current window is closed and the in-memory table’s schema is guaranteed to be updated with the DDL from the binlog stream.

<img width="1268" alt="3" src="https://user-images.githubusercontent.com/553032/138533306-f13412d3-ab5e-4d49-b4ac-8c240649a533.png">

_Since the incremental snapshot events are created during the pause in binlog processing and added into the binlog stream only after the window close event it’s possible that some binlog events preceding the chunk events have a more recent database schema than incremental snapshot events. If it’s critical to keep the schema change consistency in the binlog stream, then it can be achieved by triggering chunk re-read on DDL events in historized connectors._

jpechane commented on debezium-core/src/main/java/io/debezium/pipeline/source/snapshot/incremental/AbstractIncrementalSnapshotChangeEventSource.java:
```diff
@@ -244,18 +248,21 @@ protected void readChunk() throws InterruptedException {
+                if (connectorConfig.supportsSchemaChangesDuringIncrementalSnapshot()) {
+                    if (!context.isSchemaVerified()) {
+                        verifySchema();
+                        if (isTableInvalid()) {
```
When it is possible to have the table being invalid at this point?

kgalieva replied:
Agreed, this check is extra, removed it

jpechane commented on debezium-core/src/main/java/io/debezium/pipeline/source/snapshot/incremental/AbstractIncrementalSnapshotChangeEventSource.java:
```diff
@@ -388,6 +451,21 @@ private void createDataEventsForTable() {
+        IncrementalSnapshotSchema schema = new IncrementalSnapshotSchema(rs.getMetaData());
+        if (!schema.equals(context.getSchema())) {
+            context.setSchemaVerified(false);
+            IncrementalSnapshotSchema oldSchema = context.getSchema();
```
This variable is not used. Maybe it was intended for logging?

kgalieva replied:
Added logging

jpechane commented on debezium-core/src/main/java/io/debezium/pipeline/source/snapshot/incremental/AbstractIncrementalSnapshotContext.java:
```diff
@@ -244,6 +250,29 @@ public void maximumKey(Object[] key) {
+        this.schemaVerified = schemaVerified;
+        if (schemaVerified) {
+            LOGGER.info("Schema verified {}", schema);
+        }
```
I'd log even the other branch.

jpechane commented on debezium-core/src/main/java/io/debezium/pipeline/source/snapshot/incremental/AbstractIncrementalSnapshotContext.java:
```diff
@@ -73,6 +73,10 @@
 
+    private IncrementalSnapshotSchema schema;
+
+    private boolean schemaVerified;
```
Can we rename the flag variable. From the name it is not clear whether the verification was executed or whether the verification has passed.

jpechane commented on debezium-core/src/main/java/io/debezium/pipeline/source/snapshot/incremental/AbstractIncrementalSnapshotChangeEventSource.java:
```diff
@@ -300,6 +312,55 @@ protected void readChunk() throws InterruptedException {
+                    context.setSchemaVerified(true);
+                }
+                else {
+                    context.setSchema(currentSchema);
```
Why is the verified flag not set to `false` here?

kgalieva replied:
Refactored this method

jpechane commented on debezium-core/src/main/java/io/debezium/pipeline/source/snapshot/incremental/AbstractIncrementalSnapshotChangeEventSource.java:
```diff
@@ -300,6 +312,55 @@ protected void readChunk() throws InterruptedException {
+        return false;
+    }
+
+    private void verifySchema() throws SQLException {
```
Generally I understand the purpose and the flow of the method but it seems to me that's not easily readable for those who will go through it for the first time. Either code srestructure is needed or more comments are necessary.

kgalieva replied:
Refactored and added the comments

jpechane commented on debezium-core/src/main/java/io/debezium/pipeline/source/snapshot/incremental/AbstractIncrementalSnapshotChangeEventSource.java:
```diff
@@ -244,18 +248,21 @@ protected void readChunk() throws InterruptedException {
+                            continue;
+                        }
+                        if (!context.isSchemaVerified()) {
+                            break;
```
When this situation happens and what should be the action?

kgalieva replied:
Agreed it's confusing. Added a comment

jpechane commented on debezium-core/src/main/java/io/debezium/pipeline/source/snapshot/incremental/IncrementalSnapshotSchema.java:
```diff
@@ -0,0 +1,99 @@
+import java.util.Arrays;
+import java.util.Objects;
+
+public class IncrementalSnapshotSchema {
```
Why it is not possible to re-use standard `Table` class?

kgalieva replied:
Updated the code to use the standard Table class

jpechane commented:
@kgalieva Hi, thanks a lot, this is definitely valuable improvement! I'd like to discuss few concerns

- [ ] Could you please provide a PR also for db2 connector? If there is support for SQL Server then Db2 should work too.
- [ ]  Is not schema checking potentially expensive operation? If yes then this feature should be configurable and allowed to switch it on/off.
- [ ] I see that your schema verification does not take into consideration default value changes, PK changes. Is it intentional? If yes then the limits should be clearly documented.
- [ ] DBZ-4197 has been merged so the PR can be rebased.
- [ ] Generally I feel the code shoudl either be more documented and more logging messages (describing the transition form state to state added).

kgalieva commented:
@jpechane Thank you for the thorough review! I've refactored the code, added the config option to turn on the schema changes check with the description of the limitations. I'll look into support for db2 later, going to create a separate issue for it.

jpechane commented:
@kgalieva Excellent job, merging! I am keeping the Jira issue open as the docs updat for the new option is needed. Also please follow up with Db2 PR. This is really excellent improvement.
